// grunt-dojo2/options/clean.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    return function (grunt) {
        grunt.loadNpmTasks('grunt-contrib-clean');
        return {
            typings: {
                src: ['typings/']
            },
            dist: {
                src: ['dist/umd/*'],
                filter: function (path) {
                    return grunt.option('remove-links') ? true : !grunt.file.isLink(path);
                }
            },
            dev: {
                src: ['<%= devDirectory %>']
            },
            src: {
                src: ['{src,tests}/**/*.js'],
                filter: function (path) {
                    // Only clean the .js file if a .js.map file also exists
                    var mapPath = path + '.map';
                    if (grunt.file.exists(mapPath)) {
                        grunt.file.delete(mapPath);
                        return true;
                    }
                    return false;
                }
            },
            coverage: {
                src: ['coverage-unmapped.json']
            },
            typedoc: {
                src: ['<%= apiDocDirectory %>', '<%= apiPubDirectory %>']
            },
            ghpages: {
                src: ['<%= apiPubDirectory %>']
            }
        };
    };
});
