// grunt-dojo2/options/intern.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    return function (grunt) {
        grunt.loadNpmTasks('intern');
        return {
            options: {
                runType: 'runner',
                config: '<%= devDirectory %>/tests/intern',
                reporters: ['Runner']
            },
            browserstack: {},
            saucelabs: {
                options: {
                    config: '<%= devDirectory %>/tests/intern-saucelabs'
                }
            },
            remote: {},
            local: {
                options: {
                    config: '<%= devDirectory %>/tests/intern-local',
                }
            },
            node: {
                options: {
                    runType: 'client'
                }
            },
            proxy: {
                options: {
                    proxyOnly: true
                }
            }
        };
    };
});
