// grunt-dojo2/options/replace.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    return function (grunt) {
        grunt.loadNpmTasks('grunt-text-replace');
        return {
            addIstanbulIgnore: {
                src: ['<%= devDirectory %>/**/*.js'],
                overwrite: true,
                replacements: [
                    {
                        from: /^(var __(?:extends|decorate|param) = )/gm,
                        to: '$1/* istanbul ignore next */ '
                    },
                    {
                        from: /^(\()(function \(deps, )/m,
                        to: '$1/* istanbul ignore next */ $2'
                    }
                ]
            }
        };
    };
});
