// grunt-dojo2/options/typedoc.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    return function (_grunt) {
        return {
            options: {
                // All options but publishOptions are passed directly to the typedoc command line.
                mode: 'modules',
                externalPattern: '**/+(example|examples|node_modules|tests|typings)/**/*.ts',
                // TODO: A dummy exclude pattern is required for typedoc 0.5.6
                exclude: '_',
                excludeExternals: true,
                excludeNotExported: true,
                includeDeclarations: true,
                // publishOptions are only used when publishing the generate API docs
                publishOptions: {
                    branch: 'gh-pages',
                    deployKey: 'deploy_key',
                    subDirectory: 'api',
                    publishMode: function () {
                        // Require that the API doc deployment is explicitly requested in an environment variable
                        // this allows us to turn it off without needing to make a commit and allows forking repos
                        // to select their own settings without changing code
                        var deploy = process.env.DEPLOY_DOCS;
                        return process.env.TRAVIS_BRANCH === 'master' && deploy;
                    }
                }
            }
        };
    };
});
