// grunt-dojo2/tasks/util/postcss.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var path = require('path');
    var fs = require('fs');
    var postCssImport = require('postcss-import');
    var postCssNext = require('postcss-cssnext');
    var postCssModules = require('postcss-modules');
    var umdWrapper = require('./umdWrapper');
    function createProcessors(dest, cwd, dist) {
        if (cwd === void 0) { cwd = ''; }
        return [
            postCssImport,
            postCssNext({
                features: {
                    autoprefixer: {
                        browsers: [
                            'last 2 versions',
                            'ie >= 11'
                        ]
                    }
                }
            }),
            postCssModules({
                generateScopedName: dist ? '[hash:base64:8]' : '[name]__[local]__[hash:base64:5]',
                getJSON: function (cssFileName, json) {
                    var outputPath = path.resolve(dest, path.relative(cwd, cssFileName));
                    var newFilePath = outputPath + '.js';
                    var themeKey = ' _key';
                    json[themeKey] = 'dojo-' + path.basename(outputPath, '.m.css');
                    fs.writeFileSync(newFilePath, umdWrapper(JSON.stringify(json)));
                }
            })
        ];
    }
    exports.createProcessors = createProcessors;
    ;
});
