// grunt-dojo2/tasks/util/Publisher.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "fs", "./process", "path"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var fs_1 = require("fs");
    var process_1 = require("./process");
    var path_1 = require("path");
    var consoleLogger = {
        writeln: function (info) {
            console.log(info);
            return this;
        }
    };
    function setGlobalConfig(key, value) {
        return process_1.exec("git config --global " + key + " " + value, { silent: true });
    }
    exports.setGlobalConfig = setGlobalConfig;
    var Publisher = (function () {
        function Publisher(cloneDir, options) {
            if (options === void 0) { options = {}; }
            /**
             * The branch to publish
             */
            this.branch = 'gh-pages';
            /**
             * The deployment key to use
             */
            this.deployKey = 'deploy_key';
            /**
             * Logging utility
             */
            this.log = consoleLogger;
            this.cloneDirectory = cloneDir;
            // optional configuration values
            options.branch && (this.branch = options.branch);
            options.deployKey && (this.deployKey = options.deployKey);
            options.log && (this.log = options.log);
            if (options.url) {
                this.url = options.url;
            }
            else {
                var repo = process.env.TRAVIS_REPO_SLUG || ''; // TODO look up the repo information?
                this.url = "git@github.com:" + repo + ".git";
            }
        }
        /**
         * @return {boolean} if a deploy key exists in the file system
         */
        Publisher.prototype.hasDeployCredentials = function () {
            return fs_1.existsSync(this.deployKey);
        };
        /**
         * Commit files to a fresh clone of the repository
         */
        Publisher.prototype.commit = function () {
            if (process_1.exec('git status --porcelain', { silent: true, cwd: this.cloneDirectory }) === '') {
                this.log.writeln('Nothing changed');
                return false;
            }
            process_1.exec("git add --all .", { silent: true, cwd: this.cloneDirectory });
            process_1.exec('git commit -m "Update API docs"', { silent: true, cwd: this.cloneDirectory });
            return true;
        };
        /**
         * Initialize the repo and prepare for it to check in
         */
        Publisher.prototype.init = function () {
            var publishBranch = this.branch;
            // Prerequisites for using git
            if (!this.hasConifg('user.name')) {
                setGlobalConfig('user.name', 'Travis CI');
            }
            if (!this.hasConifg('user.email')) {
                setGlobalConfig('user.email', 'support@sitepen.com');
            }
            this.log.writeln("Cloning " + this.url);
            this.execSSHAgent('git', ['clone', this.url, this.cloneDirectory], { silent: true });
            try {
                process_1.exec("git checkout " + publishBranch, { silent: true, cwd: this.cloneDirectory });
            }
            catch (error) {
                // publish branch didn't exist, so create it
                process_1.exec("git checkout --orphan " + publishBranch, { silent: true, cwd: this.cloneDirectory });
                process_1.exec('git rm -rf .', { silent: true, cwd: this.cloneDirectory });
                this.log.writeln("Created " + publishBranch + " branch");
            }
        };
        /**
         * Publish the contents of { sourceDirectory } in the clone at { cloneDir } in the directory
         * { subDirectory } and push to the { branch } branch.
         */
        Publisher.prototype.publish = function () {
            this.log.writeln("Publishing " + this.branch + " to origin");
            this.execSSHAgent('git', ['push', 'origin', this.branch], { silent: true, cwd: this.cloneDirectory });
            this.log.writeln("Pushed " + this.branch + " to origin");
        };
        /**
         * Execute a credentialed git command
         * @param command the command to execute
         * @param options execute options
         */
        Publisher.prototype.execSSHAgent = function (command, args, options) {
            if (this.hasDeployCredentials()) {
                var deployKey = this.deployKey;
                var relativeDeployKey = options.cwd ? path_1.relative(options.cwd, deployKey) : deployKey;
                fs_1.chmodSync(deployKey, '600');
                return process_1.exec("ssh-agent bash -c 'ssh-add " + relativeDeployKey + "; " + command + " " + args.join(' ') + "'", options);
            }
            else {
                this.log.writeln("Deploy Key \"" + this.deployKey + "\" is not present. Using environment credentials.");
                var response = process_1.spawn(command, args, options);
                if (response.stderr) {
                    this.log.writeln(response.stderr);
                }
                return response.stdout;
            }
        };
        Publisher.prototype.hasConifg = function (key) {
            try {
                return !!process_1.exec("git config " + key, { silent: true });
            }
            catch (e) { }
            return false;
        };
        return Publisher;
    }());
    exports.default = Publisher;
});
