// grunt-dojo2/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "path", "glob"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var path = require("path");
    var glob = require("glob");
    function formatGlob(tsconfigGlob) {
        return tsconfigGlob.map(function (glob) {
            if (/^\.\//.test(glob)) {
                // Remove the leading './' from the glob because grunt-ts
                // sees it and thinks it needs to create a .baseDir.ts which
                // messes up the "dist" compilation
                return glob.slice(2);
            }
            return glob;
        });
    }
    exports.initConfig = function (grunt, otherOptions) {
        var tsconfigContent = grunt.file.read('tsconfig.json');
        var tsconfig = JSON.parse(tsconfigContent);
        if (tsconfig.filesGlob) {
            tsconfig.filesGlob = formatGlob(tsconfig.filesGlob);
        }
        else {
            tsconfig.include = formatGlob(tsconfig.include);
        }
        var packageJson = grunt.file.readJSON('package.json');
        var devTasks = [
            'clean:typings',
            'typings:dev',
            'tslint',
            'clean:dev',
            'copy:staticDefinitionFiles-dev',
            'dojo-ts:dev',
            'copy:staticTestFiles'
        ];
        var distTasks = [
            'clean:typings',
            'typings:dist',
            'tslint',
            'clean:dist',
            'copy:staticDefinitionFiles-dist',
            'dojo-ts:dist',
            'fixSourceMaps'
        ];
        var distESMTasks = [
            'dojo-ts:esm'
        ];
        var docTasks = [
            'clean:typings',
            'typings:dev',
            'typedoc',
            'clean:typedoc',
            'clean:ghpages'
        ];
        grunt.initConfig({
            name: packageJson.name,
            version: packageJson.version,
            tsconfig: tsconfig,
            tsconfigContent: tsconfigContent,
            filesGlob: tsconfig.filesGlob || tsconfig.include,
            all: ['<%= filesGlob %>'],
            skipTests: ['<%= all %>', '!tests/**/*.ts'],
            testsGlob: ['./tests/**/*.ts', 'tests/**/*.ts'],
            staticTestFiles: 'tests/**/*.{html,css,json,xml,js,txt}',
            staticDefinitionFiles: '**/*.d.ts',
            devDirectory: '<%= tsconfig.compilerOptions.outDir %>',
            apiDocDirectory: '_apidoc',
            apiPubDirectory: '_apipub',
            distDirectory: 'dist/umd/',
            otherOptions: otherOptions,
            devTasks: devTasks,
            distTasks: distTasks,
            distESMTasks: distESMTasks,
            docTasks: docTasks
        });
        var options = {};
        glob.sync('*.js', {
            cwd: path.join(__dirname, 'options')
        }).forEach(function (filename) {
            var optName = path.basename(filename, '.js');
            options[optName] = require('./options/' + optName)(grunt);
        });
        grunt.config.merge(options);
        require('./tasks/uploadCoverage')(grunt);
        require('./tasks/installPeerDependencies')(grunt, packageJson);
        require('./tasks/repl')(grunt, packageJson);
        require('./tasks/run')(grunt, packageJson);
        require('./tasks/release')(grunt, packageJson);
        require('./tasks/link')(grunt, packageJson);
        require('./tasks/fixSourceMaps')(grunt, packageJson);
        require('./tasks/postcss')(grunt);
        if (otherOptions) {
            grunt.config.merge(otherOptions);
        }
        require('./tasks/ts')(grunt);
        require('./tasks/typedoc')(grunt);
        // Set some Intern-specific options if specified on the command line.
        ['suites', 'functionalSuites', 'grep'].forEach(function (option) {
            var value = grunt.option(option);
            var splitValue;
            if (value) {
                if (option !== 'grep') {
                    splitValue = value.split(',').map(function (string) { return string.trim(); });
                }
                grunt.config('intern.options.' + option, splitValue || value);
            }
        });
        function setCombined(combined) {
            if (combined) {
                grunt.config('intern.options.reporters', [
                    { id: 'grunt-dojo2/lib/intern/Reporter', file: 'coverage-unmapped.json' }
                ]);
            }
        }
        setCombined(grunt.option('combined'));
        grunt.registerTask('test', (function () {
            var flags = Object.keys(this.flags);
            if (!flags.length) {
                flags.push('node');
            }
            grunt.option('force', true);
            grunt.task.run('clean:coverage');
            grunt.task.run('dev');
            setCombined(true);
            flags.forEach(function (flag) {
                grunt.task.run('intern:' + flag);
            });
            grunt.task.run('remapIstanbul:coverage');
            grunt.task.run('clean:coverage');
        }));
        grunt.registerTask('dev', grunt.config.get('devTasks'));
        grunt.registerTask('dist', grunt.config.get('distTasks'));
        grunt.registerTask('dist_esm', grunt.config.get('distESMTasks'));
        grunt.registerTask('doc', grunt.config.get('docTasks'));
        grunt.registerTask('default', ['clean', 'dev']);
    };
});
