// grunt-dojo2/tasks/postcss.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./util/postcss"], factory);
    }
})(function (require, exports) {
    "use strict";
    var postcss_1 = require("./util/postcss");
    return function init(grunt) {
        var path = require('path');
        var fs = require('fs');
        var postCssImport = require('postcss-import');
        var postCssNext = require('postcss-cssnext');
        grunt.loadNpmTasks('grunt-postcss');
        var distDirectory = grunt.config.get('distDirectory') || '';
        var devDirectory = grunt.config.get('devDirectory') || '';
        var variablesProcessors = [
            postCssImport,
            postCssNext({
                features: {
                    customProperties: {
                        preserve: 'computed'
                    }
                }
            })
        ];
        function moduleFiles(dest) {
            return [{
                    expand: true,
                    src: ['**/*.m.css'],
                    dest: dest,
                    cwd: 'src'
                }];
        }
        var cssFiles = [{
                expand: true,
                src: ['**/*.css', '!**/*.m.css'],
                dest: distDirectory,
                cwd: 'src'
            }];
        grunt.config.set('postcss', {
            options: {
                map: true
            },
            'modules-dev': {
                files: moduleFiles(path.join(devDirectory, 'src')),
                options: {
                    processors: postcss_1.createProcessors(devDirectory)
                }
            },
            'modules-dist': {
                files: moduleFiles(distDirectory),
                options: {
                    processors: postcss_1.createProcessors(distDirectory, 'src', true)
                }
            },
            variables: {
                files: cssFiles,
                options: {
                    processors: variablesProcessors
                }
            }
        });
    };
});
