// grunt-dojo2/tasks/release.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    return function (grunt, packageJson) {
        var execa = require('execa');
        var path = require('path');
        var pkgDir = require('pkg-dir');
        var parse = require('parse-git-config');
        var packagePath = pkgDir.sync(process.cwd());
        var npmBin = 'npm';
        var gitBin = 'git';
        var temp = 'temp/';
        var defaultBranch = 'master';
        var preReleaseTags = ['alpha', 'beta', 'beta1', 'beta2', 'beta3', 'rc'];
        var gitBaseRemote = 'git@github.com:dojo/';
        var defaultMaintainers = ['sitepen', 'dojotoolkit', 'dojo'];
        var extraToCopy = ['README.md'];
        var releaseVersion = grunt.option('release-version');
        var nextVersion = grunt.option('next-version');
        var preReleaseTag = grunt.option('pre-release-tag');
        var dryRun = grunt.option('dry-run');
        var tag = grunt.option('tag');
        var pushBack = grunt.option('push-back');
        var initial = grunt.option('initial');
        var skipChecks = grunt.option('skip-checks');
        var initialPackageJson = grunt.file.readJSON(path.join(packagePath, 'package.json'));
        var commitMsg = '"Update package metadata"';
        function matchesPreReleaseTag(preReleaseTag, version) {
            var regexp = new RegExp("(.*)-(" + preReleaseTag + ")\\.(\\d+)");
            return regexp.exec(version) || [];
        }
        function matchesVersion(version1, version2) {
            return version2.indexOf(version1) === 0;
        }
        function getNextPreReleaseTagVersion(versionInPackage, existingVersions) {
            var filteredVersions = existingVersions
                .filter(function (v) { return matchesVersion(versionInPackage, v); })
                .filter(function (v) { return matchesPreReleaseTag(preReleaseTag, v); })
                .map(function (v) { return parseInt(matchesPreReleaseTag(preReleaseTag, v)[3], 10) || 0; });
            var nextVersion = filteredVersions.length ? Math.max.apply(Math, filteredVersions) + 1 : 1;
            return versionInPackage + "-" + preReleaseTag + "." + nextVersion;
        }
        function preparePackageJson(packageJson) {
            packageJson.private = undefined;
            packageJson.scripts = undefined;
            packageJson.files = undefined;
            return packageJson;
        }
        function getGitRemote() {
            var gitConfig = parse.sync();
            var remotes = Object.keys(gitConfig)
                .filter(function (key) { return key.indexOf('remote') === 0; })
                .filter(function (key) { return gitConfig[key].url.indexOf(gitBaseRemote) === 0; })
                .map(function (key) { return gitConfig[key].url; });
            return remotes.length ? remotes[0] : false;
        }
        function npmPreReleaseVersion(versionInPackage, versions) {
            var versionToRelease = getNextPreReleaseTagVersion(versionInPackage, versions);
            var args = ['version', versionToRelease];
            if (dryRun) {
                args.unshift('--no-git-tag-version');
            }
            grunt.log.subhead("version to release: " + versionToRelease);
            return command(npmBin, args, {}, true);
        }
        function command(bin, args, options, executeOnDryRun) {
            if (dryRun && !executeOnDryRun) {
                grunt.log.subhead('dry-run (not running)');
            }
            bin = bin + " " + args.join(' ');
            grunt.log.ok(bin + " - " + JSON.stringify(options));
            if (!dryRun || executeOnDryRun) {
                return execa.shell(bin, options);
            }
            return Promise.resolve({});
        }
        grunt.registerTask('can-publish-check', 'check whether author can publish', function () {
            var done = this.async();
            var whoamiPromise = command(npmBin, ['whoami'], {}, true).then(function (result) { return result.stdout; }, function (err) { return grunt.fail.fatal('not logged into npm'); });
            var maintainersPromise;
            if (initial) {
                maintainersPromise = Promise.resolve(defaultMaintainers);
            }
            else {
                maintainersPromise = command(npmBin, ['view', '.', '--json'], {}, true)
                    .then(function (result) { return JSON.parse(result.stdout).maintainers; })
                    .then(function (maintainers) { return maintainers.map(function (maintainer) { return maintainer.replace(/\s<.*/, ''); }); });
            }
            return Promise.all([whoamiPromise, maintainersPromise]).then(function (results) {
                var user = results[0];
                var maintainers = results[1];
                var isMaintainer = maintainers.indexOf(user) > -1;
                if (!isMaintainer) {
                    grunt.fail.fatal("cannot publish this package with user " + user);
                }
            }).then(done);
        });
        grunt.registerTask('repo-is-clean-check', 'check whether the repo is clean', function () {
            var done = this.async();
            return command(gitBin, ['status', '--porcelain'], {}, true)
                .then(function (result) {
                if (result.stdout) {
                    grunt.fail.fatal('there are changes in the working tree');
                }
            })
                .then(function () { return command(gitBin, ['rev-parse', '--abbrev-ref', 'HEAD'], {}, true); })
                .then(function (result) {
                if (result.stdout !== defaultBranch) {
                    grunt.fail.fatal("not on " + defaultBranch + " branch");
                }
            })
                .then(done);
        });
        grunt.registerTask('release-publish', 'publish the package to npm', function () {
            var done = this.async();
            var args = ['publish', '.'];
            if (tag) {
                args.push('--tag', tag);
            }
            var promises = [command(npmBin, args, { cwd: temp }, false)];
            grunt.log.subhead('publishing to npm...');
            if (dryRun) {
                promises.push(command(npmBin, ['pack', '../' + temp], { cwd: 'dist' }, true));
            }
            return Promise.all(promises).then(done);
        });
        grunt.registerTask('release-version-pre-release-tag', 'auto version based on pre release tag', function () {
            var done = this.async();
            var versionInPackage = initialPackageJson.version.replace(/-.*/g, '');
            if (initial) {
                return npmPreReleaseVersion(versionInPackage, []).then(done);
            }
            else {
                return command(npmBin, ['view', '.', '--json'], {}, true).then(function (result) {
                    if (result.stdout) {
                        var time = JSON.parse(result.stdout).time;
                        var versions = Object.keys(time).filter(function (key) {
                            return ['created', 'modified'].indexOf(key) < 0;
                        });
                        npmPreReleaseVersion(versionInPackage, versions).then(done);
                    }
                    else {
                        grunt.fail.fatal('failed to fetch versions from npm');
                    }
                });
            }
        });
        grunt.registerTask('release-version-specific', 'set the version manually', function () {
            var done = this.async();
            var args = ['version', releaseVersion];
            if (dryRun) {
                args.unshift('--no-git-tag-version');
            }
            grunt.log.subhead("version to release: " + releaseVersion);
            return command(npmBin, args, {}, true).then(done);
        });
        grunt.registerTask('post-release-version', 'update the version post release', function () {
            var done = this.async();
            var packageJson = Object.assign({}, initialPackageJson);
            if (nextVersion) {
                packageJson.version = nextVersion;
            }
            grunt.file.write('package.json', JSON.stringify(packageJson, null, '  ') + '\n');
            grunt.log.subhead("version of package.json to commit: " + packageJson.version);
            return command(gitBin, ['commit', '-am', commitMsg], {}, false)
                .then(function () {
                if (!pushBack) {
                    return;
                }
                var remote = getGitRemote();
                if (remote) {
                    return command(gitBin, ['push', remote, defaultBranch], {}, false)
                        .then(function () { return command(gitBin, ['push', remote, '--tags'], {}, false); });
                }
                else {
                    grunt.log.subhead('could not find remote to push back to. please push with tags back to remote');
                }
            })
                .then(function () { return grunt.log.subhead('release completed. please push with tags back to remote'); })
                .then(done);
        });
        grunt.registerTask('release-publish-flat', 'publish the flat package', function () {
            grunt.log.subhead('making flat package...');
            var pkg = grunt.file.readJSON(path.join(packagePath, 'package.json'));
            var dist = grunt.config('copy.staticDefinitionFiles-dist.dest');
            var tasks = ['copy:temp', 'release-publish', 'clean:temp'];
            grunt.config.merge({
                copy: { temp: { expand: true, cwd: dist, src: '**', dest: temp } },
                clean: { temp: [temp] }
            });
            grunt.file.write(path.join(temp, 'package.json'), JSON.stringify(preparePackageJson(pkg), null, '  ') + '\n');
            extraToCopy.forEach(function (fileName) {
                if (grunt.file.exists(fileName)) {
                    grunt.file.copy(fileName, temp + '/' + fileName);
                }
            });
            grunt.task.run(tasks);
        });
        grunt.registerTask('release', 'release', function () {
            grunt.option('remove-links', true);
            var tasks = ['dist'];
            if (skipChecks && !dryRun) {
                grunt.fail.fatal('you can only skip-checks on a dry-run!');
            }
            if (!skipChecks) {
                tasks.unshift('can-publish-check', 'repo-is-clean-check');
            }
            if (preReleaseTag && preReleaseTags.indexOf(preReleaseTag) > -1) {
                tasks.push('release-version-pre-release-tag');
            }
            else if (releaseVersion && nextVersion) {
                tasks.push('release-version-specific');
            }
            else {
                grunt.fail.fatal('please specify --pre-release-tag or --release-version and --next-version');
            }
            tasks.push('release-publish-flat');
            tasks.push('post-release-version');
            grunt.task.run(tasks);
        });
    };
});
