// grunt-dojo2/tasks/ts.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "lodash"], factory);
    }
})(function (require, exports) {
    "use strict";
    var _ = require("lodash");
    return function (grunt) {
        var distDirectory = grunt.config.get('distDirectory');
        var defaultOptions = {
            dist: {
                exclude: ['tests/**/*.ts', 'src/*/tests/**/*.ts', 'src/*/example/**/*.ts'],
                compilerOptions: {
                    outDir: distDirectory,
                    declaration: true,
                    sourceMap: true,
                    inlineSources: true
                }
            },
            esm: {
                exclude: ['tests/**/*.ts', 'src/*/tests/**/*.ts', 'src/*/example/**/*.ts'],
                compilerOptions: {
                    target: 'es6',
                    module: 'es6',
                    sourceMap: false,
                    outDir: 'dist/esm',
                    inlineSourceMap: true,
                    inlineSources: true
                }
            }
        };
        grunt.registerTask('dojo-ts', function () {
            grunt.loadNpmTasks('grunt-ts');
            var flags = this.args && this.args.length ? this.args : ['dev'];
            var tsconfig = grunt.config.get('tsconfig');
            var tsOptions = grunt.config.get('ts') || {};
            var baseOptions = {
                failOnTypeErrors: true,
                fast: 'never'
            };
            grunt.config.set('ts.options', baseOptions);
            var tasks = [];
            flags.forEach(function (target) {
                var tsconfigFileName = 'tsconfig.json';
                tasks.push("ts:" + target);
                // dev task cannot be configured outside of projects tsconfig
                if (target !== 'dev') {
                    var targetTsconfig = _.cloneDeep(tsconfig);
                    var targetDefaultOptions = defaultOptions[target] || {};
                    var targetTsOptions = tsOptions[target] || {};
                    _.merge(targetTsconfig, targetDefaultOptions, targetTsOptions);
                    tsconfigFileName = ".tsconfig" + target + ".json";
                    grunt.file.write(tsconfigFileName, JSON.stringify(targetTsconfig));
                    grunt.config.set("clean." + target + "Tsconfig", { src: tsconfigFileName });
                    tasks.push("clean:" + target + "Tsconfig");
                }
                grunt.config.set("ts." + target, { tsconfig: { passThrough: true, tsconfig: tsconfigFileName } });
            });
            grunt.task.run(tasks);
        });
    };
});
